mwclient/client.py

```diff
@@ -628,7 +692,7 @@ def require(self, major, minor, revision=None, raise_error=True):
         """
         if self.version is None:
             if raise_error is None:
-                return
+                return  # type: ignore[unreachable]
```

Comment:
This is...weird. It seems like this should be reachable and the check should be `if not raise_error` rather than `if raise_error is None`. This logic seems to first appear in 465d4b0 (the early commit history is weird as Dan was messily syncing from an SVN repo maintained by Bryan, it seems), where it behaves more or less as it currently does, but there is no explicit type hint for `raise_error` except that it defaults to `True`. There's no explanation for why it uses `if raise_error is None` here, but `if raise_error` on the other branch.

In any case, the way it is right now, we would raise an error here with `raise_error=False`, which seems pretty weird. I feel like we should correct this so you can reach this outcome without violating the expected type of `raise_error`, and allow this function to return a `bool` or `None`?

Reply:
I think personally I would keep the types as is, deprecate the current behavior, and introduce a new `raise_error is False` condition. I would like to avoid complicating the types, as it would prevent simply conditions such as the following from type-checking.

```python
if site.require(1, 27, raise_error=False):
    ...
```

Instead we could do something like the following, keeping the old behavior but having it fail type-checks and logging a `DeprecationWarning`.


```diff
diff --git a/mwclient/client.py b/mwclient/client.py
index 8f192ff..92e8be8 100644
--- a/mwclient/client.py
+++ b/mwclient/client.py
@@ -692,7 +692,14 @@ class Site:
         """
         if self.version is None:
             if raise_error is None:
-                return  # type: ignore[unreachable]
+                warnings.warn(  # type: ignore[unreachable]
+                    'Passing raise_error=None to require is deprecated and will be '
+                    'removed in a future version. Use raise_error=False instead.',
+                    DeprecationWarning
+                )
+                return
+            elif raise_error is False:
+                return False
             # FIXME: Replace this with a specific error
             raise RuntimeError(f'Site {repr(self)} has not yet been initialized')
```

Reply:
But that doesn't address the weird situation that calling it with `raise_error=False` before the site is initialized will raise an error, the thing you specifically told it not to do...

Reply:
Alright, then let's change the return type to `Optional[bool]` after all. I was wrong about this preventing calls such as `if site.require(1, 27, raise_error=False)`, as `None` is falsy. I'll still add the `DeprecationWarning` to the `raise_error is None` case, as that is just straight up _weird_.